pkg/appmanager/eventNotifier.go

```diff
@@ -0,0 +1,133 @@
+/*-
+ * Copyright (c) 2016,2017, F5 Networks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package appmanager
+
+import (
+	"strings"
+	"sync"
+
+	log "github.com/F5Networks/k8s-bigip-ctlr/pkg/vlogger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/tools/record"
+)
+
+type (
+	NamespaceEventNotifierMap map[string]*NamespaceEventNotifier
+
+	NewBroadcasterFunc func() record.EventBroadcaster
+
+	EventNotifier struct {
+		mutex           sync.Mutex
+		notifierMap     map[string]*NamespaceEventNotifier
+		broadcasterFunc NewBroadcasterFunc
+	}
+
+	NamespaceEventNotifier struct {
+		broadcaster record.EventBroadcaster
+		recorder    record.EventRecorder
+	}
+)
+
+func NewEventNotifier(bfunc NewBroadcasterFunc) *EventNotifier {
+	if nil == bfunc {
+		// No broadcaster func provided (unit testing), use real one.
+		bfunc = record.NewBroadcaster
+	}
+	return &EventNotifier{
+		notifierMap:     make(map[string]*NamespaceEventNotifier),
+		broadcasterFunc: bfunc,
+	}
+}
+
+func (en *EventNotifier) getNotifierForNamespace(
```

Comment:
I think this should be two functions, one to get and one to create. It does more than it says it does which could confuse a caller.

Reply:
I can rename it and create a second version for the unit test case that's simpler, but I don't want 2 functions for the normal usage path because then there's 2 locks/unlocks and a chance for something to happen in between. The get (or create and get) should occur with a single lock/unlock.